File: 2024/day01.py
def task1(xs: list[int], ys: list[int]) -> int:
    return sum(abs(x - y) for (x, y) in zip(xs, ys))


def task2(xs: list[int], ys: list[int]) -> int:
    freq = {y: ys.count(y) for y in ys}
    return sum(x * freq.get(x, 0) for x in xs)


def main():
    with open("input.txt") as f:
        lines = [[int(n) for n in line.split()] for line in f]
        xs, ys = map(sorted, zip(*lines))
        
        print(task1(xs, ys))
        print(task2(xs, ys))


if __name__ == "__main__":
    main()

File: 2022/src/days/day01.rs
use std::{cmp::Reverse, error::Error};

fn task_1(calories: &[u32]) -> u32 {
    *calories.first().expect("Missing first calorie")
}

fn task_2(calories: &[u32]) -> u32 {
    calories.iter().take(3).sum()
}

pub fn main(input: String) -> Result<(), Box<dyn Error>> {
    let mut calories: Vec<u32> = input
        .split("\n\n")
        .map(|group| group.lines().map(|line| line.parse::<u32>().unwrap()).sum())
        .collect();
    calories.sort_by_key(|c| Reverse(*c));
    println!("task 1: {}", task_1(&calories));
    println!("task 2: {}", task_2(&calories));
    Ok(())
}
